lms/djangoapps/certificates/signals.py

```diff
@@ -81,17 +82,38 @@ def _listen_for_id_verification_status_changed(sender, user, **kwargs):  # pylin
 
 def fire_ungenerated_certificate_task(user, course_key, expected_verification_status=None):
     """
-    Helper function to fire un-generated certificate tasks
-
-    The 'mode_is_verified' query is copied from the GeneratedCertificate model,
-    but is done here in an attempt to reduce traffic to the workers.
-    If the learner is verified and their cert has the 'unverified' status,
-    we regenerate the cert.
+    Helper function to fire certificate generation task.
+    Auto-generation of certificates is available for following course modes:
+        1- VERIFIED
+        2- CREDIT_MODE
+        3- PROFESSIONAL
+        4- NO_ID_PROFESSIONAL_MODE
+
+    NOTE: Purpose of restricting other course modes (HONOR and AUDIT) from auto-generation is to reduce
+    traffic to workers.
     """
+
+    # course enrollment modes allowed for auto-generation of certificates
+    allowed_enrollement_modes_list = [
+        CourseMode.VERIFIED,
+        CourseMode.CREDIT_MODE,
+        CourseMode.PROFESSIONAL,
+        CourseMode.NO_ID_PROFESSIONAL_MODE,
+    ]
     enrollment_mode, __ = CourseEnrollment.enrollment_mode_for_user(user, course_key)
-    mode_is_verified = enrollment_mode in GeneratedCertificate.VERIFIED_CERTS_MODES
     cert = GeneratedCertificate.certificate_for_student(user, course_key)
-    if mode_is_verified and (cert is None or cert.status == 'unverified'):
+
+    if cert is None:
+        generate_learner_certificate = (enrollment_mode in allowed_enrollement_modes_list)
+    else:
```

Comment:
We regenerate a certificate when its status is `unverified`. A certificate have `unverified` status when a user has one of `GeneratedCertificate.VERIFIED_CERTS_MODES` enrollment and is not verified, see https://github.com/edx/edx-platform/blob/master/lms/djangoapps/certificates/queue.py#L287-L292. We need to run certificate generation task so that certificate status can be changed to `verified` if user has since verified.
So, We are only checking a regeneration possibility in cases where enrollment is from `GeneratedCertificate.VERIFIED_CERTS_MODES`.